typing/value_rec_check.ml

```diff
@@ -189,28 +208,130 @@ let classify_expression : Typedtree.expression -> sd =
+          | Record_extension _ ->
+              Static (Regular_block (1 + (Array.length fields)))
+          | Record_unboxed _ -> assert false (* handled above *)
+        end
     | Texp_tuple exprs ->
-        if List.for_all is_constant exprs then Constant else Static
+        if List.for_all is_constant exprs
+        then Constant
+        else Static (Regular_block (List.length exprs))
 
     | Texp_apply ({exp_desc = Texp_ident (_, _, vd)}, _)
       when is_ref vd ->
-        Static
+        Static (Regular_block 1)
     | Texp_apply (_,args)
       when List.exists is_abstracted_arg args ->
-        Static
+        Static Function
     | Texp_apply _ ->
         Not_recursive
 
-    | Texp_new _
-    | Texp_instvar _
-    | Texp_array _
-    | Texp_pack _
-    | Texp_object _
-    | Texp_function _
-    | Texp_lazy _
+    | Texp_array fields ->
+        begin match Typeopt.array_kind e with
+        | Lambda.Pfloatarray ->
+          Static (Float_record (List.length fields))
+        | Lambda.Pgenarray | Lambda.Paddrarray | Lambda.Pintarray ->
+          Static (Regular_block (List.length fields))
+        end
+    | Texp_pack mexp ->
+        let size_of_structure (s : Typedtree.structure) =
+          let rec size_of_structure_items acc items =
+            match (items : Typedtree.structure_item list) with
+            | [] -> acc
+            | item :: rem ->
+              let acc =
+                match item.str_desc with
+                | Tstr_value (_rec_flag, pat_expr_list) ->
+                  acc + List.length (let_bound_idents pat_expr_list)
+                | Tstr_typext tyext ->
+                  acc + List.length tyext.tyext_constructors
+                | Tstr_exception _ext ->
+                  acc + 1
+                | Tstr_module ({mb_presence=Mp_present} as mb) ->
+                  begin match mb.mb_id with
+                  | None -> acc
+                  | Some _ -> acc + 1
+                  end
+                | Tstr_recmodule bindings ->
+                  let ids = (List.filter_map (fun mb -> mb.mb_id) bindings) in
+                  acc + List.length ids
+                | Tstr_class cl_list ->
+                  acc + List.length cl_list
+                | Tstr_include incl ->
+                  let ids = bound_value_identifiers incl.incl_type in
+                  acc + List.length ids
+                | Tstr_open od ->
+                  let ids = bound_value_identifiers od.open_bound_items in
+                  acc + List.length ids
+                | Tstr_primitive _
+                | Tstr_type _
+                | Tstr_eval _
+                | Tstr_module ({mb_presence=Mp_absent})
+                | Tstr_modtype _
+                | Tstr_class_type _
+                | Tstr_attribute _ ->
+                  acc
+            in
+            size_of_structure_items acc rem
+          in
+          size_of_structure_items 0 s.str_items
+        in
+        let rec classify_mexp mexp : sd =
```

Comment:
This case is a mouthful. Could you move `classify_mexp` as a mutually-recursive `classify_module_expression` function, with `size_of_structure` a local function of `classify_module_expression`?

Reply:
Done